### ISSUE TITLE
Meta service adds tags twice (once for / and once for the current route)

### ISSUE THREAD

**Issue by shreyagarwaldev** [angular/angular#19606]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[x] Bug report  
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
</code></pre>

## Current behavior
I am using Meta service right now in an Angular 4 Universal application. So I add a few meta tags on the home component (route /) and the same ones with different values on another page (some route /abc). Now if I go to route /, i only see the meta tags for that page, but if i go to any other route, I see the meta tags for route / followed by css styles, followed by meta tags for the page I am currently on. This is on server side and client side both.

I have tried using forceCreation, updateTags and every other thing possible. Unable to solve. 

## Expected behavior
Meta tags should only be shown for current page.

## Minimal reproduction of the problem with instructions
Try this out on https://www.thepixelatedplanet.com. View page source for the landing page and then navigate to workshops and view page source for that. See the 2 blocks of meta tags with different values.

## What is the motivation / use case for changing the behavior?
Since the meta tags for the / route appear first, they override the actual page's meta tags. This breaks previews and other such functionality

## Environment
@angular/cli: 1.4.3
node: 6.11.1
os: win32 x64
@angular/animations: 4.4.3
@angular/common: 4.4.3
@angular/compiler: 4.4.3
@angular/core: 4.4.3
@angular/forms: 4.4.3
@angular/http: 4.4.3
@angular/platform-browser: 4.4.3
@angular/platform-browser-dynamic: 4.4.3
@angular/platform-server: 4.4.3
@angular/router: 4.4.3
@angular/cli: 1.4.3
@angular/compiler-cli: 4.4.3
@angular/language-service: 4.4.3
typescript: 2.3.4

Browser:
- [x] Chrome (desktop) version XX
- [x] Chrome (Android) version XX
- [x] Chrome (iOS) version XX
- [x] Firefox version XX
- [x] Safari (desktop) version XX
- [x] Safari (iOS) version XX
- [x] IE version XX
- [x] Edge version XX
 
For Tooling issues:
- Node version: node: 6.11.1
- Platform:  Mac, Windows

Others:
Please provide a workaround if not a fix.

**Comment by Toxicable** [comment]
Hello, we reviewed this issue and determined that it doesn't fall into the bug report or feature request category. This issue tracker is not suitable for support requests, please repost your issue on [StackOverflow](http://stackoverflow.com/) using tag `angular`.

If you are wondering why we don't resolve support issues via the issue tracker, please [check out this explanation](https://github.com/angular/angular/blob/master/CONTRIBUTING.md#-got-a-question-or-problem).

**Comment by ocombe** [comment]
Actually I think that it is, since the meta service is from angular, why do you think that it's not an actual bug report for us @Toxicable ?

**Comment by Toxicable** [comment]
>Now if I go to route /, i only see the meta tags for that page, but if i go to any other route, I see the meta tags for route / followed by css styles, followed by meta tags for the page I am currently on.

This indicates that the issue is caused by how the Meta service is being used rather than a bug with the service itself. If there is a bug the Meta service then to verify it a plunkr/Stack Blitz would be needed.

**Comment by ocombe** [comment]
Hard to do a plunkr for an universal app unfortunately :(

**Comment by Toxicable** [comment]
>This is on server side and client side both.

I assume that's meaning independently.

Side note: Stack Blitz is actually working on supporting Node apps at sometime Q1 next year

**Comment by Agrumas** [comment]
This can be related to #19093 try to remove `BrowserAnimationsModule`

**Comment by Toxicable** [comment]
@Agrumas I think that is unlikely. The Meta service is almost directly editing the head tags' DOM, it's never interacting with views, which is where your issue linked resides

**Comment by shreyagarwaldev** [comment]
@Toxicable I have followed online examples for meta service and the code is straightforward enough to not cause this issue. I even tried remove tag and force creation to get rid of the twice occurring meta tags but it didn't work. I've never used plunkr or stackblitz but will try to give it a shot. As I said, you can work the issue at www.thepixelatedplanet.com and I can provide cover from there.

**Comment by shreyagarwaldev** [comment]
@Toxicable I have followed online examples for meta service and the code is straightforward enough to not cause this issue. I even tried remove tag and force creation to get rid of the twice occurring meta tags but it didn't work. I've never used plunkr or stackblitz but will try to give it a shot. As I said, you can see the issue at www.thepixelatedplanet.com and I can provide code from there.

**Comment by Toxicable** [comment]
Seeing the result, the linked website, is not reasonable enough for us to action this issue.
We cannot see the source code, we cannot identify if it's a problem with your app or with Angular itself.
This is why we require a plunkr ot Stack Blitz, there is just too many moving parts in a full application.

**Comment by shreyagarwaldev** [comment]
@Toxicable those platforms don't support universal. I checked and this only shows for server rendered html. I can give you the simplest repro in a github.com repo hosted somewhere if that works?

**Comment by shreyagarwaldev** [comment]
@Toxicable @ocombe @Agrumas  I have a super simple universal app with this repro now.
Code --> https://github.com/shreyagarwaldev/MetaBugDemo
Demo --> https://piplanet-shrey.herokuapp.com/ 
There are 2 pages, super empty. Just try viewing source for each. One of them will repeat the meta tags. I only have 2 meta tags (og:title, twitter:title). And when it repeats, the first value is from the other page.

This is an issues with server side rendering. I am not necessarily sure how else I can convince you guys that there is an issue here :) I hoped it was something in my app, but it seems to be a legit bug.

**Comment by shreyagarwaldev** [comment]
tl/dr - RESOLVED WITH WORKAROUND. This is a bug in how prerender works. 

Well, I went ahead and investigated and found the problem. The server side rendered view for index.html (the one used for template engine) already has the <meta> tags from the <app-root> while it should only have those meta tags which are explicitly in the index.html. This happens for any element created in the head region. The workaround was to change the prerender code to remove these extra meta tags as follows. This also repros with Renderer. So its not just a Meta Service issue but a deeper issue.

Look at the html.replace part. That's the workaround.

renderModuleFactory(AppServerModuleNgFactory, {
  document: index,
  url: '/',
  extraProviders: [
    provideModuleMap(LAZY_MODULE_MAP)
  ]
})
.then(html => writeFileSync(join('browser', 'index.html'), html.replace(new RegExp('<meta name="[a-z]+:[a-z_]+" content="[A-Za-z0-9 ,\'!:/\. -]+">', 'g'), '').replace(new RegExp('<meta property="[a-z]+:[a-z_]+" content="[A-Za-z0-9,\'!:/\. -]+">', 'g'), '').replace(new RegExp('<meta name="[a-z]+" content="[A-Za-z0-9,\'!:/\. -]+">', 'g'), '')));

**Comment by magnattic** [comment]
Same problem for me, it works

**Comment by magnattic** [comment]
We have the same problem. `Meta#updateTag()` works fine without Universal, but as soon as we add Universal the tags are rendered twice, one with the default value from index.html and one with for the updated one.

@shreyagarwaldev Can you repost your workaround? It looks like some of the code got swallowed by github?

**Comment by magnattic** [comment]
Any news on this? Does anyone have a workaround?

A big use case of Angular Universal is SEO, so if the meta tags are not working properly in Universal this is kind of a big issue...

**Comment by alamgirqazi** [comment]
> Any news on this? Does anyone have a workaround?
> 
> A big use case of Angular Universal is SEO, so if the meta tags are not working properly in Universal this is kind of a big issue...
> 
> **Update**: It actually was an unrelated bug in our code, we accidentally used
> `this.meta.updateTag({ name: 'og:description', content: description });`
> instead of
> `this.meta.updateTag({ property: 'og:description', content: description });`
> which lead to the duplicated tags during runtime.

Thank you!

**Comment by adrobson** [comment]
I am using addTags([{.....}]) and it was duplicating the last element. However, after my last element I had a comma. When I removed the comma after the last element it worked correctly and the last tag wasn't duplicated.

**Comment by petebacondarwin** [comment]
OK, so I see what is happening here...

When the Meta service is trying to work out if the meta already exists it searches for the first meta element that has the same name as the one that is being added. It then checks to see if this element also matches the other attributes. If not it adds a new one.

In this case the first meta is always found and it doesn't match the new (second) meta that is being added so a new element is always created.

Instead we should be using `querySelectorAll()` to get all the meta elements that match the meta name, and then filtering down those, only adding a new one if the filtered list is empty.
